### ISSUE TITLE
Feat: Support for tagging/grouping systems

### ISSUE THREAD

**Issue by Muska-Ami** [henrygd/beszel#87]
I switched from Nezha Monitoring to Beszel, mainly because Beszel supports search, while Nezha Monitoring does not. In this switch, I seem to have lost an important feature: adding tags/groups to Systems. I think if this feature is added, it will make the monitoring project more intuitive.
Could you please add this feature? Thanks.

**Comment by henrygd** [comment]
Yes, I'll add this. Probably not in the next release but hopefully this month.

**Comment by iesope** [comment]
Yes please!
Perhaps this feature might help someone like myself that has dockers spread between many LXC's in proxmox

**Comment by rozpuszczalny** [comment]
I second that! One could e.g. add a docker label with a list of tags and later allow to group by tags (`beszel.tags: tag1,tag2`)

**Comment by Glitch3dPenguin** [comment]
I also think that adding some form of "sites", "groups", or tagging feature would be really helpful. Things get a little crazy when I want to add 10+ servers. I really love what you are building here! Thanks!

**Comment by henrygd** [comment]
Sorry, I've put this off for a while because I couldn't decide how to implement it.

@dakshpareek even [asked a few days ago](https://github.com/henrygd/beszel/issues/333#issuecomment-2580136394) if he could start working on this, and I asked him to hold off until I had a clear idea of what to do.

I'm going to add an 'edit system' dialog soon (similar to the add system dialog), so that should be a good place to edit tags / groups.

I'm mostly wondering what the best method is to group systems or display tags. Or if just hooking the tags up to the search and view options is enough.

I want to research how other software handles it, but please let me know if you have opinions or examples!

**Comment by emigrating** [comment]
For me, it would be enough to simply allow tags for each system. Then display all systems like now but with an added choice of grouping the display based on tags.

I've only just started installing Beszel on the various services I run, and I already have over 30 systems. Currently, I'm naming the systems like **[TYPE] Name**, for instance:

- **[media] voicesampler**
- **[web] vaultwarden**
- **[management] pve-dl-380**
- **[services] OPS-1**

This approach works, but it's still fairly cumbersome. Since this system only really allows for a single 'tag,' I'm limited to grouping them based on service type.

If I could attach several tags to each system, I could group them based on the type of service (media, web, services, etc.) and the actual server they are running on. This would be useful if a service shows higher-than-nominal bandwidth or disk I/O. I could quickly estimate whether it was an issue with the VM host rather than the service itself.

**Comment by bmcgonag** [comment]
First, thank you so very much for this amazing software. I truly appreciate it, and it's been a great tool. 

I know you've been struggling with how to implement this, and while I can't give a technical answer on that, I'm happy to suggest a few ways from the user's perspective that it could be done.  Both from assigning groups / tags, as well as actually displaying machines in groups / by tag.

### User Entry
#### Option 1
Allow the user to add groups / tags directly on the main table view of the dashboard.  Add a column for tags / groups, and when clicked allow the user to type, providing auto-type suggestions that have already been used and save to the database. The user, however, should be able to create a completely new tag in this way as well.

#### Option 2 (my personal favorite as a single user)
Add an action on the Action column (3-dot) called 'Assign Groups', or 'Tag'. When clicked provide a small modal allowing the user to add groups / tags (as chips perhaps?), again showing type-ahead options for things that exist, but still allowing the user to create new groups / tags by simply typing and entering them, then clicking 'Save' to dismiss the modal. 

#### Option 3 (something I would imagine multi-user systems would benefit from)
Make group creation part of the overall settings, and tie it to a Role / Permission. This would give an administrator the ability to define groups / tags, and the ability to assign a role to their users to only be able to use those groups / tags that have already been created. Next, use either of Option 1 or 2 above for the end user to add groups / tags to systems if they have the permissions to do so.

### Displaying Groups of Systems
#### Option 1
Use collapsible sections (one per group / tag), and when expanded show the systems that belong in that group / have that tag.  Have a default 'All' that would display all systems, but allow a full admin to disable theall group if they want to in the settings.

#### Option 2 (my personal favorite idea)
Add a filter above the table which filters the full list of systems by the tag(s) / group(s) the user types into the filter.  This should be a filter that has type-ahead, and is limited to show and allow only existing tags / groups.  Use a chip style to allow mutiple to be typed.  Then provide a Check Box to select 'Match Any' or 'Match All'.  If 'Match Any' then if a system has any of the entered groups or tags, but not all of them it will still display. If 'Match All' then a system needs to belong to all groups or have all tags in order to display.

#### Option 3
Add a Left side panel slide-out with various filtering options in it, including Groups / Tags, and provide the ability to multi-select groups / tags in the slide out.  

These are just a few ideas, but happy to help with designs as wellif you'd like them. Seems like you have a pretty good handle on layout and views already.

**Comment by imilne** [comment]
Just chipping in with some comments, and want to mention it's worth taking a look at ganglia. It's pretty old now, but still usable for 1000+ node systems because of the way you can both create groups (or clusters) of hosts *and* get summary overviews of each group. It's nice to see things like my docker hosts are using 40% CPU on average, so I should drill down and see what's up there, or that the network bandwidth across the compute nodes is very high, so let's have a look at what's happening, etc.

**Comment by antunnitraj** [comment]
bump

**Comment by imclint21** [comment]
+1 — this gets painful fast at scale. Once you cross a few hundred systems (let alone thousands), the flat list is basically unusable even with search.

For me the minimal useful version is just:

- a `tag` (or key) field on each system
- a "group by tag" toggle on the dashboard that collapses systems under each tag

That's it. One click, everything organized, no need to scroll through hundreds of rows. No need to overthink groups vs tags vs roles/permissions up front — you can layer richer aggregation (à la ganglia) on top later without breaking the data model.

**Comment by trtiger4520** [comment]
I would love to see this feature include user association/permissions!

Right now, managing user access machine-by-machine gets tedious at scale. It would be amazing if we could assign groups or tags to specific users, so they automatically get access to all systems under that tag. This would save a lot of time instead of having to configure each machine individually.

**Comment by thutex** [comment]
came here to see if this was on the roadmap, because indeed, some sort of grouping so you can display only systems within a group would be a useful feature

**Comment by Soul-Code** [comment]
I’m in dire need of it.

**Comment by 0x5t4l1n** [comment]
The maintainer mentioned being unsure about the data model — here's one way to think about it that keeps things flexible without over-engineering:

**A flat tag model (not hierarchical groups) with multi-select filtering** is probably the right starting point. Reasons:
- It doesn't force users to choose between "groups" vs "tags" — tags *are* groups when you filter by one
- Multi-tag assignment means one system can appear in "production" AND "europe" AND "nginx" without duplicating it
- The PocketBase schema is a simple string array column with an index — no relational complexity

For the UX, one comment above mentioned Ganglia's cluster-level aggregation. That's compelling but it's a separate feature from basic grouping. I'd suggest splitting it:
- **v1**: tag assignment in the edit dialog + tag-based filter chips above the system table (like the existing search, but stickier)
- **v2**: aggregate stats view per tag (useful for "all my Proxmox nodes" as a rolled-up row)

On the user-permission tie-in that came up: tying tags to access control would require significant auth refactoring. Worth scoping that separately so it doesn't block v1 — most single-operator deployments just need visibility grouping, not access isolation.

Would a tag-per-system (no inheritance, no nesting) with a multi-select filter cover your top use cases?

**Comment by gmckeown** [comment]
My own requirement here is very simple; I have some servers that are always online and some servers that are only brought online when required... so I would love to be able to filter the server list to exclude the expected offline ones.

There seems to be a bit of analysis paralysis on this issue. The ability to add tags to servers would be useful from a number of perspectives, and then grouping, filtering, etc. could be built on top of that over time?

**Comment by KyleMercer** [comment]
I think @0x5t4l1n and @gmckeown have both managed to hit it on the head. Not looking for anything too complicated here. My use case is to use one tag for always online vs part-time similar to @gmckeown; then to have another tag for the type of system (PC, Docker, Proxmox, etc). A flat model is acceptable here for sure.

No inheritance and no nesting is needed here for my use case nor is user-permission.
